pyvista/core/utilities/fileio.py

```diff
@@ -1056,26 +1056,37 @@ def to_meshio(mesh: DataSet) -> meshio.Mesh:
     connectivity = mesh.cell_connectivity
 
     # Generate polyhedral cell faces if any
-    polyhedral_cells = pyvista.convert_array(mesh.GetFaces())
+    if pyvista.vtk_version_info < (9, 4):
+        polyhedral_cells: VectorLike[int] = pyvista.convert_array(mesh.GetFaces())
 
-    if polyhedral_cells is not None:
-        locations = pyvista.convert_array(mesh.GetFaceLocations())
-        polyhedral_cell_faces = []
+        if polyhedral_cells is not None:
+            locations = pyvista.convert_array(mesh.GetFaceLocations())
+            polyhedral_cell_faces: list[list[VectorLike[int]]] = []
 
-        for location in locations:
-            if location == -1:
-                continue
+            for location in locations:
+                if location == -1:
+                    polyhedral_cell_faces.append([])
+                    continue
+
+                n_faces = polyhedral_cells[location]
+                i = location + 1
+                faces: list[VectorLike[int]] = []
 
-            n_faces = polyhedral_cells[location]
-            i = location + 1
-            faces: list[MatrixLike[int]] = []
+                while len(faces) < n_faces:
+                    n_vertices = polyhedral_cells[i]
+                    faces.append(polyhedral_cells[i + 1 : i + 1 + n_vertices])
+                    i += n_vertices + 1
 
-            while len(faces) < n_faces:
-                n_vertices = polyhedral_cells[i]
-                faces.append(polyhedral_cells[i + 1 : i + 1 + n_vertices])
-                i += n_vertices + 1
+                polyhedral_cell_faces.append(faces)
 
-            polyhedral_cell_faces.append(faces)
+    else:
+        from ..cell import _get_irregular_cells
+
+        polyhedron_faces = _get_irregular_cells(mesh.GetPolyhedronFaces())
+        polyhedron_locations = _get_irregular_cells(mesh.GetPolyhedronFaceLocations())
```

Comment:
Should we try rebasing this PR to `maint/vtk-override` from https://github.com/pyvista/pyvista/pull/6879? `polyhedron_faces` and `polyhedron_face_locations` properties are added there.

```suggestion
        polyhedron_faces = mesh.polyhedron_faces
        polyhedron_locations = mesh.polyhedron_face_locations
```

Reply:
I don't think these two are equivalent, though. The one with ``_get_irregular_cells`` returns a list of faces for each cell, while I think the latter returns a 1D array. Anyway, I think it would be better to directly implements these small changes in the VTK 9.4 PR and close this one.

Reply:
> I don't think these two are equivalent, though. The one with `_get_irregular_cells` returns a list of faces for each cell, while I think the latter returns a 1D array. 

Let's keep your current implementation then.

> Anyway, I think it would be better to directly implements these small changes in the VTK 9.4 PR and close this one.

I am fine with either merging this PR with main (after merging #6879) _or_ we rebase this PR to that one so that we can merge this directly into #6897 _before_  it gets merged into main.

There are some "not yet implemented for vtk 9.4" warnings currently in #6879 that would be nice to remove:
https://github.com/pyvista/pyvista/blob/556751bb8573ee88acbffe4067b8cee42b29a6ac/pyvista/core/utilities/fileio.py#L1078-L1080

but at the same time there may also be a slight preference to sort this stuff out later, e.g. see https://github.com/pyvista/pyvista/pull/6879#issuecomment-2764249651

Due to some of the complexity involved with rebasing, perhaps it's best to do the latter i.e. leave #6879 as-is, and resolve any merge conflicts and fix the vtk 9.4 polyhedron stuff only after that PR lands.

Reply:
I implemented ``polyhedron_faces`` and ``polyhedron_face_locations`` for VTK < 9.4, and used these two new attributes in ``to_meshio()``. Looks cleaner and more maintainable.